File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root and not subRoot:
            return True
        
        if not root or not subRoot:
            return False

        def match (node, subnode):
            if not node and not subnode:
                return True
            
            if not node or not subnode:
                return False

            if node.val != subnode.val:
                return False
            
            if not match(node.left, subnode.left) or not match(node.right, subnode.right):
                return False

            return True
        
        if root.val == subRoot.val and match(root, subRoot):
            return True
    
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> ret;
    vector<int> rightSideView(TreeNode* root) {
        int level = 0;
        ret = fillNumber(root, level);
        return ret;
    }

private:
    vector<int> fillNumber(TreeNode* root, int level) {
        if (!root) {
            return ret;
        }

        if (ret.size() == level) {
            ret.push_back(root -> val);
        }

        fillNumber(root -> right, level + 1);
        fillNumber(root -> left, level + 1);

        return ret;
    }
};
